Data/Primitive/Array.hs

```diff
@@ -521,17 +521,20 @@ instance Alternative Array where
 instance Monad Array where
   return = pure
   (>>) = (*>)
-  a >>= f = push 0 [] (sizeofArray a - 1)
+
+  ary >>= f = collect 0 [] (la-1)
    where
-   push !sz bs i
-     | i < 0 = build sz bs
-     | otherwise = let b = f $ indexArray a i
-                    in push (sz + sizeofArray b) (b:bs) (i+1)
-
-   build sz stk = createArray sz (die ">>=" "impossible") $ \mb ->
-     let go off (b:bs) = copyArray mb off b 0 (sizeofArray b) >> go (off + sizeofArray b) bs
-         go _   [    ] = return ()
-      in go 0 stk
+   la = sizeofArray ary
+   collect sz stk i
+     | i < 0 = createArray sz (die ">>=" "impossible") $ fill 0 stk
+     | otherwise = let sb = f $ indexArray ary i in
+         collect (sz + sizeofArray sb) (sb:stk) (i-1)
```

Comment:
Is the key part of fixing the infinite loop the use of a minus sign here instead of a plus sign? (It's not obvious at a quick glance.)

Reply:
I never really dug into that bit. The `SmallArray` code seemed to work, do I copied it.